File: master/longest_substring_without_repeating_char.py
class Solution:
    # @return an integer
    def lengthOfLongestSubstring(self, s):
        if len(s)<2:
            return len(s)
        m = [0]*256
        start = end = 0
        res = 0
        while end<len(s):
            if m[ord(s[end])] == 0:
                res = max(res,end-start+1)
                m[ord(s[end])]+=1
                end += 1
                continue
            while True:
                m[ord(s[start])] -= 1
                start += 1
                if s[start-1] == s[end]:
                    break
        return res
                

print Solution().lengthOfLongestSubstring("abcabcbb")

File: 5/longest_substring_without_repeating_char.cpp
#include "header.h"

int lengthOfLongestSubstring(string s) {
	if(s.length()<2) return s.length();
	int sz = s.length();
	int map[127] = {0};
	int l = 0, r=1;
	map[s[l]] = 1;
	int res = 1;
	while(r<sz){
		if(map[s[r]]==0){
			map[s[r]]++;
			r++;
			res = max(res,r-l);
			continue;
		}
		while(map[s[r]]>0){
			map[s[l]]--;
			l++;
		}
		map[s[r]]=1;
		res = max(res,r-l+1);
		r++;
	}
	return res;
}	

int main(){
    printf("%d\n", lengthOfLongestSubstring("abcabcbb")); //abc
    printf("%d\n", lengthOfLongestSubstring("bbbbb")); //b
    return 0;
}
